source/common/tcp_proxy/tcp_proxy.h

```diff
@@ -113,21 +113,43 @@ using RouteConstSharedPtr = std::shared_ptr<const Route>;
 using TunnelingConfig =
     envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_TunnelingConfig;
 
+/**
+ * Base class for both tunnel response headers and trailers.
+ */
+class TunnelResponseHeadersOrTrailers : public StreamInfo::FilterState::Object {
+public:
+  ProtobufTypes::MessagePtr serializeAsProto() const override;
+  virtual const Http::HeaderMap& value() const PURE;
+};
+
 /**
  * Response headers for the tunneling connections.
  */
-class TunnelResponseHeaders : public StreamInfo::FilterState::Object {
+class TunnelResponseHeaders : public TunnelResponseHeadersOrTrailers {
 public:
   TunnelResponseHeaders(Http::ResponseHeaderMapPtr&& response_headers)
       : response_headers_(std::move(response_headers)) {}
-  const Http::ResponseHeaderMap& value() const { return *response_headers_; }
-  ProtobufTypes::MessagePtr serializeAsProto() const override;
+  const Http::ResponseHeaderMap& value() const override { return *response_headers_; }
```

Comment:
```suggestion
  const Http::HeaderMap& value() const override { return *response_headers_; }
```

Reply:
Can you, please, explain why to do this change?
Also what you suggest is a breaking change for those are using "value()" with the return type "ResponseHeaderMap&".

Reply:
@yanavlasov ??

Reply:
While it is legal in C++ to override a virtual function with a different return type as long as they are covariant, it is not really needed in this case. You can safely return the reference to the base type of the header map.

Reply:
This previous code:
https://github.com/envoyproxy/envoy/blob/f6cb005211c389df0dc17d71b6819912e083b5cd/source/common/tcp_proxy/tcp_proxy.h#L123

Reply:
I've changed the method's signature in order to use it in "serializeAsProto" method:
https://github.com/envoyproxy/envoy/blob/b6bd4eacc63082bd2e7aa7e2991e0bac2eaf912c/source/common/tcp_proxy/tcp_proxy.cc#L567
And to keep the previous use of the method in the ResponseHeaderMap class by returning  a covariant type.
 
However, if you are not ok with that, I can do the following: 

- Remove the declaration of "value()" from the base class (TunnelResponseHeadersOrTrailers) and keep it as a regular method in both derived classes for those who need to get a return value of ResponseHeaderMap/ResponseTrailerMap (In my case, I need the derived class specific type).

- Introduce another method in the base class (let's say "virtual const Http::HeaderMap& headerMap() const PURE;") for the use in "serializeAsProto" method.

WDYT?

Reply:
The value() method was unused (I can not find any existing uses for it). You can change its return value as you need.

Reply:
It's not used here in the repo, but it is could be used by not built-in filters (I have a filter that uses this method with the current return type).

Reply:
I see. If you are using the `value()` method in your proprietary code, when yes, you need to create a new virtual method.

Also note, that the `value()` method may be deleted as unused code in the future, since maintainers have no way to know that a proprietary Envoy's extension is using it. To avoid this, you need to create tests that use this method and add comment stating that it is also used in a proprietary code to avoid its deletion.

Reply:
Got it, thanks!